agent/taskresource/asmauth/asmauth_test.go

```diff
@@ -0,0 +1,143 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package asmauth
+
+import (
+	"encoding/json"
+	"testing"
+	"time"
+
+	apicontainer "github.com/aws/amazon-ecs-agent/agent/api/container"
+	apitaskstatus "github.com/aws/amazon-ecs-agent/agent/api/task/status"
+	"github.com/aws/amazon-ecs-agent/agent/asm"
+	"github.com/aws/amazon-ecs-agent/agent/asm/factory/mocks"
+	mock_secretsmanageriface "github.com/aws/amazon-ecs-agent/agent/asm/mocks"
+	"github.com/aws/amazon-ecs-agent/agent/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/credentials/mocks"
+	"github.com/aws/amazon-ecs-agent/agent/taskresource"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/secretsmanager"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	executionCredentialsID = "exec-creds-id"
+	region                 = "us-west-2"
+	secretID               = "meaning-of-life"
+	username               = "irene"
+	password               = "sher"
+	taskARN                = "task1"
+)
+
+var (
+	asmAuthDataVal       string
+	requiredASMResources []*apicontainer.ASMAuthData
+)
+
+func init() {
+	asmAuthDataBytes, _ := json.Marshal(&asm.AuthDataValue{
+		Username: aws.String(username),
+		Password: aws.String(password),
+	})
+	asmAuthDataVal = string(asmAuthDataBytes)
+	requiredASMResources = []*apicontainer.ASMAuthData{
+		{
+			CredentialsParameter: secretID,
+			Region:               region,
+		},
+	}
+}
+
+func TestCreateAndGet(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	credentialsManager := mock_credentials.NewMockManager(ctrl)
+	asmClientCreator := mock_factory.NewMockClientCreator(ctrl)
+	mockASMClient := mock_secretsmanageriface.NewMockSecretsManagerAPI(ctrl)
+
+	iamRoleCreds := credentials.IAMRoleCredentials{}
+	creds := credentials.TaskIAMRoleCredentials{
+		IAMRoleCredentials: iamRoleCreds,
+	}
+	asmSecretValue := &secretsmanager.GetSecretValueOutput{
+		SecretString: aws.String(asmAuthDataVal),
+	}
+	gomock.InOrder(
+		credentialsManager.EXPECT().GetTaskCredentials(executionCredentialsID).Return(creds, true),
+		asmClientCreator.EXPECT().NewASMClient(region, iamRoleCreds).Return(mockASMClient),
```

Comment:
I don't have a solution for you here, but we need to be careful with how many layers we go especially with new code. Remember:

> Every time a mock returns a mock a fairy dies [*](https://static.javadoc.io/org.mockito/mockito-core/2.7.9/org/mockito/Mockito.html#RETURNS_DEEP_STUBS)

Reply:
yea ended up adding this to be able to test more complete workflows like post unmarshall or the image pull workflows.